### PR TITLE
Fix to #35024 - Query could not be translated when using a static ICollection/IList field

### LINKED ISSUE #35024: Query could not be translated when using a static ICollection/IList field
- Efcore seems to be unable to translate a `static readonly ICollection<T>` & `static readonly IList<T>` field into a query when using it inside of a `.Contains()` When removing the `static` keyword it seems to be able to execute the query.
- When using `IEnumerable<T>` This is not an issue, is this because of the different `.Contains()` that's being used?
- This is only happening when the List/Collection is static.
- If the collection is a static Property instead of a static field this will not throw this exception.
- Using the private static field as a local variable before using it inside of the query will not throw this exception.
- Using an `.Any(x => x.Equals(...)` also does not throw this exception.
## Code example:

```
public class Item
{
    public string Bar { get; set; }
    public string Foo { get; set; }
}
```

In my class I have the following field (This can be an `static readonly IList<T>` or `static readonly ICollection<T>`:

```
private static readonly IList<string> AllowedFooCodes = new List<string>() { "ABC", "CED" };
```

My Repository class looks something like this:

```
public class SampleDataRepository : ISampleDataRepository
{
    private readonly ILogger<SampleDataRepository> _logger;
    private readonly SampleDataStore _sampleDataStore;

    // Can be IList or ICollection, exception will be thrown regardless.
    private static readonly IList<string> AllowedFooCodes = new List<string>() { "ABC", "CED" };
    
    public SampleDataRepository(SampleDataStore store, ILogger<SampleDataRepository> logger)
    {
        _sampleDataStore = store;
        _logger = logger;
    }
    
    public async Task<IEnumerable<Item>> GetItems()
    {
        _sampleDataStore.Items
                        .AsNoTracking()
                        .Where(item => AllowedFooCodes.Contains(item.foo))
                        .ToListAsync();
    }
 
}
```

StackTrace:

```
The LINQ expression 'DbSet<Item>()
	.Where(d => (IList<string>)List<string> { "ABC", "CED" }
        .Contains(item.foo) could not be translated. Additional information: Translation of method 'System.Linq.Enumerable.Contains' failed. If this method can be mapped to your custom function, see https://go.microsoft.com/fwlink/?linkid=2132413 for more information. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator() in System.Runtime.CompilerServices\ConfiguredCancelableAsyncEnumerable.cs:line 61
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.<ToListAsync>d__67`1.MoveNext()
```

EF Core version: 8.0.3.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0
Operating system: 10.0.19045.0
IDE: Visual Studio 2022 17.9.6

### PR DESCRIPTION
**Pull request by maumar**
Problem was that when converting primitive collection to inline query root we were not matching the expression if it was constant wrapped in convert. Fix is to remove convert for the purpose of pattern match for the transformation.

Fixes #35024

### CODE REVIEW

**Comment by ChrisJollyAU** [comment]
@maumar That looks fine.

I did though ask myself where the `Convert` was actually coming in from.

Turns out it gets added as part of the funcletizer https://github.com/dotnet/efcore/blob/507e7f196a11aebaf26d1cf34e5d6a3e7044bb52/src/EFCore/Query/Internal/ExpressionTreeFuncletizer.cs#L1941

```
return constantExpression.Type != returnType
            ? Convert(constantExpression, returnType)
            : constantExpression;
```

LH Type is ```List`1``` and RH Type is ```IList`1```

If you want to prevent the convert you could go something like this
```
return constantExpression.Type != returnType && !constantExpression.Type.GetInterfaces().Contains(returnType)`
```

**Comment by maumar** [comment]
@ChrisJollyAU I'm bit apprehensive about adding (more) reflection to the funcletizer as it is a perf-sensitive component. We'd also need to cover the case of DeriveType : BaseType (e.g. ObservableCollection), so that's some more reflection. Although it would fix the problem at the very source. @roji any thoughts on this?

**Comment by ChrisJollyAU** [comment]
As a side not just FYI. 

The `IEnumerable` case does also get the `Convert` added at the same point. Its just that in the `QueryableMethodNormalizingExpressionVisitor`, in the `VistMethodCall` function it can get handled. With `IEnumerable` it calls `TryConvertEnumerableToQueryable` which strips the convert away. If you have a `IList` or `ICollection` (and it is a `Contains`) it calls  `TryConvertCollectionContainsToQueryableContains`. This does not strip the same type of convert

File: test/EFCore.Specification.Tests/Query/NorthwindAggregateOperatorsQueryTestBase.cs [hunk 1]
```diff
@@ -1361,6 +1361,15 @@ public virtual Task Contains_with_local_anonymous_type_array_closure(bool async)
             ss => ss.Set<OrderDetail>().Where(o => ids.Contains(new { Id1 = o.OrderID, Id2 = o.ProductID })));
     }
 
+    private static readonly IList<string> StaticIds = new List<string> { "ALFKI", "ANATR" };
+
+    [ConditionalTheory]
+    [MemberData(nameof(IsAsyncData))]
+    public virtual Task Contains_with_static_IList(bool async)
```
Comment (roji) [inline test/EFCore.Specification.Tests/Query/NorthwindAggregateOperatorsQueryTestBase.cs:1368]: Is there a specific reason this is in this test class (there doesn't seem to be anything aggregate about the query).

Maybe PrimitiveCollectionsQueryTestBase is a better place...

**Review by roji — COMMENTED** [review]
This seems like a correct fix for the specific issue, but I think the general behavior of constantizing "init-only" ([code](https://github.com/dotnet/efcore/blob/main/src/EFCore/Query/Internal/ExpressionTreeFuncletizer.cs#L854)) is problematic... The fact that a (static) member is init-only (readonly) doesn't mean that it's immutable; for example, in the example of the test you just wrote, there's nothing preventing the contents StaticIds from being mutated (even if StaticIds itself is readonly). But the way we currently translate, the contents of the list are captured and cached the first time the query is compiled (because they're constantized).

In other words, I thing we have an unrelated bug here; if that bug is fixed (by no longer constantizing init-only fields), then we'd no longer have the Convert node that is also causing the trouble here.

(note that if users really do want to constantize, they can still do that by introducing EF.Constant()).

What do you think?

**Comment by roji** [comment]
> I'm bit apprehensive about adding (more) reflection to the funcletizer as it is a perf-sensitive component. We'd also need to cover the case of DeriveType : BaseType (e.g. ObservableCollection), so that's some more reflection. Although it would fix the problem at the very source. @roji any thoughts on this?

Aside from my above comment, @ChrisJollyAU's suggestion makes sense to me - I think simply checking `constantExpression.Type.IsAssignableTo(returnType)` should be sufficient, and take care of interfaces/subclasses/etc. I think perf here should be OK - the funcletizer is indeed perf-sensitive, but not to the point where we can't allow this kind of check where it's needed (remember we use the LINQ interpreter just below to evaluate parameters, which is much slower). So I'd do this regardless of the removal of init-only constantization (discussed above).

**Comment by ChrisJollyAU** [comment]
> This seems like a correct fix for the specific issue, but I think the general behavior of constantizing "init-only" ([code](https://github.com/dotnet/efcore/blob/main/src/EFCore/Query/Internal/ExpressionTreeFuncletizer.cs#L854)) is problematic... The fact that a (static) member is init-only (readonly) doesn't mean that it's immutable; for example, in the example of the test you just wrote, there's nothing preventing the contents StaticIds from being mutated (even if StaticIds itself is readonly). But the way we currently translate, the contents of the list are captured and cached the first time the query is compiled (because they're constantized).

I did experiment with changing that to be `containsCapturedVariable: true`. It mostly works with the values as a parameter instead of a constant but there are a couple of interesting tests.

Indexof_with_emptystring
```
SELECT [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE CASE
    WHEN [c].[Region] IS NOT NULL THEN 0
END = 0
```

new sql
```
@__Empty_0='' (Size = 15)

SELECT [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE CHARINDEX(@__Empty_0, [c].[Region]) - CASE
    WHEN @__Empty_0 = N'' THEN 0
    ELSE 1
END = 0
```

Note that we seem to lose the optimization for that depends on a literal empty string constant. `string.Empty` is `static readonly`

Contains_on_readonly_enumerable
```
SELECT [w].[Id], [w].[AmmunitionType], [w].[IsAutomatic], [w].[Name], [w].[OwnerFullName], [w].[SynergyWithId]
FROM [Weapons] AS [w]
WHERE [w].[AmmunitionType] = 1
```

New sql
```
@___weaponTypes_0='[1]' (Size = 4000)

SELECT [w].[Id], [w].[AmmunitionType], [w].[IsAutomatic], [w].[Name], [w].[OwnerFullName], [w].[SynergyWithId]
FROM [Weapons] AS [w]
WHERE [w].[AmmunitionType] IN (
    SELECT [w0].[value]
    FROM OPENJSON(@___weaponTypes_0) WITH ([value] int '$') AS [w0]
)
```

Array_cast_to_IEnumerable_Contains_with_constant
```
SELECT [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE [c].[CustomerID] IN (N'ALFKI', N'WRONG')
```

New sql
```
@___customers_0='["ALFKI","WRONG"]' (Size = 4000)

SELECT [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE [c].[CustomerID] IN (
    SELECT [c0].[value]
    FROM OPENJSON(@___customers_0) WITH ([value] nchar(5) '$') AS [c0]
)
```

Specially in the above cases it probably makes more sense to constantize.

It also breaks `SpatialSqlServerTest.Translators_handle_static_members` with `Microsoft.Data.SqlClient.SqlException : The incoming tabular data stream (TDS) remote procedure call (RPC) protocol stream is incorrect. Parameter 1 ("@__Empty_0"): The supplied value is not a valid instance of data type geography. Check the source data for invalid values. An example of an invalid value is data of numeric type with scale greater than precision.`

**Comment by roji** [comment]
@ChrisJollyAU thanks for taking a look; it definitely makes sense that some optimizations would be lost due to moving from constantization to parameterization. The problem here is that we can't reliably know when the type in question is fully immutable, which is when it's safe to constantize. We could have a hard-coded list of the common immutable types (string, int, DateTime...) and constantize for those (when they're init-only of course), but constantizing for a type list `List<T>` (as in the example above) simply seems like a bug...

**Comment by ChrisJollyAU** [comment]
Off the top of my head, I think the only ones with the problem are those like List, Collection, Queue, HashSet, Dictionary i.e. those that implement IList and ICollection.

But definitely for your basic types that you know are completely immutable (like a readonly string or int) my thoughts are towards keeping the constantizing

> but constantizing for a type list `List<T>` (as in the example above) simply seems like a bug

Is that because of the caching or something like that (because its not in the actual sql thats generated as that works)

**Comment by roji** [comment]
> But definitely for your basic types that you know are completely immutable (like a readonly string or int) my thoughts are towards keeping the constantizing

Yeah, agree.

>> but constantizing for a type list List<T> (as in the example above) simply seems like a bug
>
> Is that because of the caching or something like that (because its not in the actual sql thats generated as that works)

Apologies - I thought a little bit about this, and I was wrong above.

The problem with constantization of a mutable thing (e.g. `List<T>`), is that if that list is then actually mutated, we get a cache miss - the funcletizer outputs a different tree. This means that the query gets fully recompiled every time the mutable thing has a different actual value: the recompilation is quite expensive, going through the entire query pipeline. We're basically missing the advantage of having parameters in the first place: that the same query compilation artifacts can be reused without recompilation even though the (parameterized) values change across executions.

So I don't think it's quite a bug, more like a problematic performance problem.

File: test/EFCore.Specification.Tests/Query/NorthwindAggregateOperatorsQueryTestBase.cs (hunk 1 above)
Comment (maumar) [reply test/EFCore.Specification.Tests/Query/NorthwindAggregateOperatorsQueryTestBase.cs:1368]: there is a lot of Contains tests in that class, so just keeping them all together. I don't mind moving them all to PCQTB as I do agree it's better place for them.

Comment (roji) [reply test/EFCore.Specification.Tests/Query/NorthwindAggregateOperatorsQueryTestBase.cs:1368]: Not very important obviously... Whatever you prefer
